src/main/CommandHandler.cpp

```diff
@@ -568,22 +568,28 @@ CommandHandler::tx(std::string const& params, std::string& retStr)
         {
             // add it to our current set
             // and make sure it is valid
-            TransactionQueue::AddResult status =
-                mApp.getHerder().recvTransaction(transaction);
+            int64_t feeBumpMinFee;
+            auto status =
+                mApp.getHerder().recvTransaction(transaction, feeBumpMinFee);
+            output << "{"
+                   << "\"status\": "
+                   << "\"" << TX_STATUS_STRING[static_cast<int>(status)]
+                   << "\"";
 
-            if (status == TransactionQueue::AddResult::ADD_STATUS_PENDING)
+            switch (status)
+            {
+            case TransactionQueue::AddStatus::ADD_STATUS_PENDING:
             {
                 StellarMessage msg;
                 msg.type(TRANSACTION);
                 msg.transaction() = envelope;
                 mApp.getOverlayManager().broadcastMessage(msg);
             }
-
-            output << "{"
-                   << "\"status\": "
-                   << "\"" << TX_STATUS_STRING[static_cast<int>(status)]
-                   << "\"";
-            if (status == TransactionQueue::AddResult::ADD_STATUS_ERROR)
+            break;
+            case TransactionQueue::AddStatus::ADD_STATUS_BAD_REPLACE_BY_FEE:
```

Comment:
Sorry, I should have noticed this much earlier. I just realized something that impacts the overall structure of this PR.

I don't think we need to add a new internal status result... or even a `minFee` element in the json:

what we really want here is that if the transaction fails with `txINSUFFICIENT_FEE`, we provide the "mininum fee" required for that transaction.

For replace by fee, it's the computation that you've done, for others, it depends on context (grep for `tx`).

OK, so if we want to always return a min fee with this kind of error... how do we return it?

For that, I suspect we can simply use `feeCharged` (in the result) to propagate the min fee on error from anywhere in the code (NB: it may already be the case for the other failures), instead of having special handling for fee bump failures coming from the mempool.

It's probably safe to do as fees are always charged before applying transactions (true now and for all older protocol versions).

Having this included in the result is actually quite nice if we can do it because I imagine that we already propagate back the raw result to the SDKs so the changes to Horizon & clients should be minimal.

Reply:
I don’t really think this is a good idea. First, I think it would be unnecessary. If the fee is insufficient for any reason other than replace-by-fee backoff then you can just calculate the required fee using the base fee from the ledger header. In that case, stellar-core doesn’t need to give you any new information. 

Second, I think there is already a huge amount of confusion regarding the results of transaction submission and validation. Your proposal definitely wouldn’t help reduce that confusion, and it could make it worse. For example, people might wonder why stellar-core is charging a fee higher than the fee specified in the transaction (it isn’t, of course, it’s just telling you what fee would have been required). This is just a single way this could be misinterpreted. But good interfaces should be easy to use correctly and hard to use incorrectly. The interface as described in @2opremio PR is obvious and unambiguous:

- You only get that result if you submit a fee-bump (this was part of your motivation of making replace-by-fee only happen for fee-bumps) so there are no backwards compatibility issues
- The result clearly indicates what happened and cannot be misinterpreted

If you really wanted to always provide fee information whenever a transaction is rejected because the fee is bad, I would actually argue that we should add `ADD_STATUS_BAD_FEE` instead of `ADD_STATUS_BAD_REPLACE_BY_FEE` and provide information there. But again, I think this is unnecessary.

Reply:
There are a few things here:
* the main point is that this proposal is "replace by fee" centric, and is not future proof.
* on the contract front
  * we are already returning a transaction result, and `feeCharged` is already set to some arbitrary thing. If we were to actually formally define what the value of `feeCharged` is on failure, I think it would be a net improvement over the current situation where people don't realize that it's the same.
  * that said, if we were to add a new field on failure because we're not happy with transaction result, we should just make the interface harder to use wrong (and it would not have two "fees" returned in one call). I had a proposal to remove using transaction results altogether for this kind of failure but there was not much appetite for having to support a different contract from the Horizon team (I think because it involves changing SDKs). This change requires changing SDKs etc - so maybe it's the right time.

Reply:
@2opremio After looking at the transaction result contract more carefully, I think the approach mentioned by @MonsieurNicolas is a good idea. It'll be a lot less code, and downstream systems will already be wired up for it. Long run, I still think we should be making this interface less confusing as noted in #1999. Sorry about the change of plans.